benchmarks/scripts/plotter.py

```diff
@@ -0,0 +1,303 @@
+
+  # Filter methods that do not have the gpu size
+  for method, df in dataframes.items():
+    if df[df['gpus'] == int(fixed_gpu_size[0])].empty:
+      continue
+  # filter gpu size not contained in the data
+  fixed_gpu_size = [gpu for gpu in fixed_gpu_size if not dataframes[list(dataframes.keys())[0]][dataframes[list(dataframes.keys())[0]]['gpus'] == int(gpu)].empty]
+  # Calculate the number of subplots based on the number of GPU sizes
+  num_subplots = len(fixed_gpu_size)
+  # Calculate the number of rows and columns for the subplots
+  num_rows = int(np.ceil(np.sqrt(num_subplots)))
+  num_cols = int(np.ceil(num_subplots / num_rows))
+  
+  # Create the subplots
+  fig, axs = plt.subplots(num_rows, num_cols, figsize=figure_size)
+  if num_subplots > 1:
+    axs = axs.flatten()  # Flatten the axs array for easier indexing
+  else:
+    axs = [axs]
+  # Plot the data for each GPU size
+  for i, gpu_size in enumerate(fixed_gpu_size):
+    ax = axs[i]  # Select the current subplot
+    
+    data_sizes = []
+    
+    # Plot the data for each method
+    for method, df in dataframes.items():
+      # Filter for the chosen number of GPUs
+      df = df[df['gpus'] == int(gpu_size)]
+      if df.empty:
+        continue
+      # Sort by data size
+      df = df.sort_values(by=['x'])
+      data_sizes.extend(df['x'].values)
+      
+      # Filter by backend and plot one line for each backend
+      if simple_plot.get(method) is not None:
+        ax.plot(df['x'].values, df['time'], marker='o', linestyle='-', label=f"{method}")
+        continue
+      
+      for backend in ['MPI', 'NCCL']:
+        df_backend = df[df['backend'] == backend]
+        
+        if df_backend.empty:
+          continue
+        
+        # Get the fastest decomposition (number of GPUs and nodes must be fixed here)
+        df_decomp = df_backend.groupby(['x', 'y', 'z', 'backend','nodes'])
+        all_decomp = []
+        sorted_dfs = []
+        
+        for _, group in df_decomp:
+          group.sort_values(by=['time'], inplace=True, ascending=False)
+          for px, py in zip(group['px'].values, group['py'].values):
+            all_decomp.append(f"{px}x{py}")
+          sorted_dfs.append(group.iloc[0])
+        
+        sorted_df = pd.DataFrame(sorted_dfs)
+        label = f"{method}-{backend}-{group['nodes'].values[0]}nodes" if nodes_in_label else f"{method}-{backend}"
+        
+        ax.plot(sorted_df['x'].values, sorted_df['time'], marker='o', linestyle='-', label=label)
+    
+    # add title nb of gpus
+    ax.set_title(f"Number of GPUs: {gpu_size}")
+    # Set x-axis ticks exactly at the provided data size values
+    ax.set_yscale('log')
+    data_sizes = list(dict.fromkeys(data_sizes))
+    data_sizes.sort()
+    ax.set_xticks(data_sizes)
+    ax.set_yticks([1e-3, 1e-2, 1e-1, 1e0])
+    
+    # Set labels and title
+    ax.set_xlabel('Data Size')
+    ax.set_ylabel('Time')
```

Comment:
FYI if you are not aware yet pandas has its own plotting API on top of matplotlib so you can do the same without all the burden once your dataframe is ready
https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html